File: MobileAppReactNativeMobX/src/business/state/CashBalance.ts
import { makeAutoObservable } from 'mobx';
import { print, round } from '../utils/utils';

class CashBalance {
  public amount: number;

  constructor(amount: number) {
    if (isNaN(amount)) this.amount = 0;
    else this.amount = round(amount);

    makeAutoObservable(this);
  }

  public setAmount(amount: number): void {
    this.amount = round(amount);
  }

  public add(howMuch: number): void {
    this.amount = round(this.amount + howMuch);
    print('Cash balance is now: ' + this.amount);
  }

  public remove(howMuch: number): void {
    this.amount = round(this.amount - howMuch);
    if (this.amount < 0) this.amount = 0;
    print('Cash balance is now: ' + this.amount);
  }

  public toString(): string {
    return `US$ ${this.amount.toFixed(2)}`;
  }

  public equals(other: CashBalance): boolean {
    return (this === other) || (this.amount === other.amount);
  }
}

export default CashBalance;

File: MobileAppFlutterRedux/lib/business/state/cash_balance.dart
import 'package:flutter/material.dart';

import '../utils/utils.dart';

@immutable
class CashBalance {
  static const CashBalance ZERO = CashBalance._(0);

  final double amount;

  CashBalance(double amount) : amount = amount.isNaN ? 0 : round(amount);

  const CashBalance._(this.amount);

  CashBalance withAmount(double amount) {
    return CashBalance(round(amount));
  }

  CashBalance add(double howMuch) {
    double newAmount = round(amount + howMuch);
    print('Added $howMuch. Cash balance is now: $newAmount.');
    return CashBalance(newAmount);
  }

  CashBalance remove(double howMuch) {
    double newAmount = round(amount - howMuch);
    if (newAmount < 0) newAmount = 0;
    print('Removed $howMuch. Cash balance is now: $newAmount.');
    return CashBalance(newAmount);
  }

  @override
  String toString() {
    return 'US\$ ${amount.toStringAsFixed(2)}';
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is CashBalance && runtimeType == other.runtimeType && amount == other.amount;

  @override
  int get hashCode => amount.hashCode;
}
